Branching stages

bonahona opened the pull request:
Added new worker function branchIf(), onTransientFailure() and onPermanentFailure().

bonahona committed:
Added test branchIf worker function.

bonahona committed:
Added test cases for branchIf worker function.
Fixed bugs in how it worked.

bonahona committed:
Updates comment for branchIf worker function.

bonahona committed:
Added new variants of branchIf to take a predicate callback instead of just a bool.
Added possibility to send messages to branchIf. Is displayed if the branch is entered.

bonahona committed:
Added typealias to predicate.
Changed end() to allow for messages.

bonahona committed:
Added new worker functions onTransientFailure and onPermanentFailure.

bonahona committed:
Fixed return type of processOnFailure

bonahona committed:
Added test cases for new onTransientFailure and onPermanentFailure worker functions.

williamhogman commented on src/main/kotlin/JobState.kt:
```diff
@@ -2,6 +2,10 @@ package franz
 
 import mu.KotlinLogging
 
+typealias Predicate<U> = suspend (U) -> Boolean
```
I like this.

williamhogman commented on src/main/kotlin/JobState.kt:
```diff
@@ -48,8 +52,17 @@ class JobState<U: Any> constructor(val value: U?, val interceptors: List<WorkerI
+    suspend fun advanceIf(msg: String, predicate: Predicate<U>): JobState<U> = processPredicate(JobStatus.Success, predicate, msg)
+
+    /**
+     * Use this when you wan't to branch of the execution of the worker. When the predicate is true, a now job state
```
s/wan't/want

williamhogman commented on src/main/kotlin/JobState.kt:
```diff
@@ -48,8 +52,17 @@ class JobState<U: Any> constructor(val value: U?, val interceptors: List<WorkerI
+
+    /**
+     * Use this when you wan't to branch of the execution of the worker. When the predicate is true, a now job state
+     * worker s created and executed to either an end() or jobStatus().
```
"is created"

"an end() or jobStatus() is called."

williamhogman commented on src/main/kotlin/JobState.kt:
```diff
@@ -69,23 +82,22 @@ class JobState<U: Any> constructor(val value: U?, val interceptors: List<WorkerI
         log.info { "Ended with status ${this.status.name}" }
         return status
     }
+    */
```
Remove, we can restore it with Git.

williamhogman reviewed (changes requested):
Minor docs changes

bonahona committed:
Fixed typos in comments.
Reworded comments regarding branchIf.
Removed old deprecated end() method implementation.
